File: apps/wearos/app/src/main/java/com/togplay/vaimarcia/wear/playback/PlaybackStrategyResolver.kt
package com.togplay.vaimarcia.wear.playback

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothA2dp
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.content.pm.PackageManager
import com.togplay.vaimarcia.wear.data.companion.WearCompanionClient
import com.togplay.vaimarcia.wear.domain.AudioCodec
import com.togplay.vaimarcia.wear.domain.DeviceCapabilityProfile
import com.togplay.vaimarcia.wear.domain.PlaybackStrategy
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.suspendCancellableCoroutine
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

/**
 * Implementa o algoritmo de decisão de docs/DEVICE_DETECTION.md §"Algoritmo" passo 3.
 * Resolvido uma vez por evento de inicialização / mudança de pareamento e mantido em cache —
 * nunca no caminho de toque-até-som (veja docs/ARCHITECTURE.md §4).
 */
@Singleton
class PlaybackStrategyResolver @Inject constructor(
    @ApplicationContext private val context: Context,
    private val companionClient: WearCompanionClient,
) {

    suspend fun resolve(): PlaybackStrategy {
        val profile = probeLocalCapabilities()
        return decide(profile)
    }

    internal fun decide(profile: DeviceCapabilityProfile): PlaybackStrategy = when {
        profile.hasBluetoothClassicAudio || profile.hasBleAudioSupport ->
            if (profile.canPlayArbitraryLocalAudio) PlaybackStrategy.DIRECT else PlaybackStrategy.RELAY
        profile.hasPersistentCompanionChannel -> PlaybackStrategy.RELAY
        else -> PlaybackStrategy.PHONE_ONLY
    }

    /**
     * docs/DEVICE_DETECTION.md passo 2 — sondagem Wear OS: `BluetoothAdapter.getProfileProxy(A2DP)`
     * mais verificação de que um dispositivo está conectado diretamente a *este* relógio (não
     * via relay pelo celular), e `PackageManager.hasSystemFeature(FEATURE_AUDIO_OUTPUT)` para
     * capacidade de reprodução de áudio local arbitrário.
     */
    private suspend fun probeLocalCapabilities(): DeviceCapabilityProfile {
        val adapter = BluetoothAdapter.getDefaultAdapter()
        val hasA2dpConnectedDevice = adapter?.let { probeA2dpConnected(it) } ?: false
        val canPlayLocalAudio = context.packageManager.hasSystemFeature(PackageManager.FEATURE_AUDIO_OUTPUT)

        return DeviceCapabilityProfile(
            manufacturer = android.os.Build.MANUFACTURER,
            model = android.os.Build.MODEL,
            osVersion = android.os.Build.VERSION.RELEASE,
            hasBluetoothClassicAudio = hasA2dpConnectedDevice,
            hasBleAudioSupport = false, // Detecção de LE Audio precisa de BluetoothLeAudioCodecConfig, API 33+; TODO
            supportedCodecs = setOf(AudioCodec.SBC, AudioCodec.AAC),
            canPlayArbitraryLocalAudio = canPlayLocalAudio,
            hasPersistentCompanionChannel = companionClient.isPhoneReachable(),
            estimatedLatencyMs = if (hasA2dpConnectedDevice) 60 else 120,
        )
    }

    private suspend fun probeA2dpConnected(adapter: BluetoothAdapter): Boolean =
        suspendCancellableCoroutine { continuation ->
            adapter.getProfileProxy(
                context,
                object : BluetoothProfile.ServiceListener {
                    override fun onServiceConnected(profile: Int, proxy: BluetoothProfile) {
                        val connected = (proxy as? BluetoothA2dp)?.connectedDevices?.isNotEmpty() ?: false
                        adapter.closeProfileProxy(BluetoothProfile.A2DP, proxy)
                        if (continuation.isActive) continuation.resume(connected)
                    }

                    override fun onServiceDisconnected(profile: Int) {
                        if (continuation.isActive) continuation.resume(false)
                    }
                },
                BluetoothProfile.A2DP,
            )
        }
}

File: apps/watchos/VaiMarciaWatch Watch App/Playback/PlaybackStrategyResolver.swift
import AVFoundation
import Foundation
import WatchKit

/// Implementa o algoritmo de decisão de docs/DEVICE_DETECTION.md §"Algoritmo" passo 3.
/// Resolvido uma vez por evento de inicialização / mudança de pareamento e mantido em cache —
/// nunca no caminho de toque-até-som (veja docs/ARCHITECTURE.md §4).
final class PlaybackStrategyResolver {
    private let connectivity: WatchConnectivityManager

    init(connectivity: WatchConnectivityManager = .shared) {
        self.connectivity = connectivity
    }

    func resolve() -> PlaybackStrategy {
        decide(probeLocalCapabilities())
    }

    func decide(_ profile: DeviceCapabilityProfile) -> PlaybackStrategy {
        if profile.hasBluetoothClassicAudio || profile.hasBleAudioSupport {
            return profile.canPlayArbitraryLocalAudio ? .direct : .relay
        }
        if profile.hasPersistentCompanionChannel {
            return .relay
        }
        return .phoneOnly
    }

    /// docs/DEVICE_DETECTION.md passo 2 — sondagem watchOS: `AVAudioSession.currentRoute`
    /// mais verificação de uma saída Bluetooth diretamente pareada; watchOS ≥ 9 em Apple
    /// Watch Ultra/Series 8+ com áudio Bluetooth direto reporta `hasBluetoothClassicAudio = true`.
    private func probeLocalCapabilities() -> DeviceCapabilityProfile {
        let session = AVAudioSession.sharedInstance()
        let hasDirectBluetoothRoute = session.currentRoute.outputs.contains { output in
            output.portType == .bluetoothA2DP || output.portType == .bluetoothLE
        }
        let supportsDirectHardware = watchSupportsDirectBluetoothAudio()

        return DeviceCapabilityProfile(
            model: WKInterfaceDevice.current().model,
            osVersion: WKInterfaceDevice.current().systemVersion,
            hasBluetoothClassicAudio: hasDirectBluetoothRoute && supportsDirectHardware,
            hasBleAudioSupport: false, // Detecção de LC3/LE Audio precisa de dados específicos do dispositivo ainda não presentes na tabela de compatibilidade
            supportedCodecs: [.aac, .sbc],
            canPlayArbitraryLocalAudio: true, // AVAudioPlayer reproduz arquivos locais arbitrários em todas as versões de watchOS que suportamos
            hasPersistentCompanionChannel: connectivity.isReachable,
            estimatedLatencyMs: supportsDirectHardware ? 70 : 130,
        )
    }

    /// Lista de permissão estática conforme a tabela de compatibilidade de
    /// docs/DEVICE_DETECTION.md: o watchOS não expõe uma API em tempo de execução para
    /// perguntar "este hardware tem rádio de áudio Bluetooth Classic", então identificar
    /// Ultra/Series 8+ vs SE/≤7 depende de comparar com identificadores de modelo conhecidos,
    /// atualizados conforme a Apple lança novas gerações de relógio.
    private func watchSupportsDirectBluetoothAudio() -> Bool {
        let identifier = watchModelIdentifier()
        return DirectAudioCapableModels.identifiers.contains(identifier)
    }

    private func watchModelIdentifier() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let mirror = Mirror(reflecting: systemInfo.machine)
        let identifier = mirror.children.reduce("") { partial, element in
            guard let value = element.value as? Int8, value != 0 else { return partial }
            return partial + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }
}

/// Watch5,x = Series 8, Watch6,x = famílias da classe Ultra/Series 9 no esquema de
/// identificadores da Apple; mantido como uma pequena tabela estática em vez de uma consulta
/// de capacidade via SDK, conforme docs/DEVICE_DETECTION.md (o SO não expõe "tem rádio de
/// áudio Bluetooth direto").
private enum DirectAudioCapableModels {
    static let identifiers: Set<String> = [
        "Watch6,1", "Watch6,2", "Watch6,3", "Watch6,4", // Series 8
        "Watch6,6", "Watch6,7", "Watch6,8", "Watch6,9", // Ultra
        "Watch7,1", "Watch7,2", "Watch7,3", "Watch7,4", // Series 9
        "Watch7,5", // Ultra 2
    ]
}
